lib/install/server/dispatcher/buffered/dispatcher.go

```diff
@@ -0,0 +1,121 @@
+/*
+Copyright 2019 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// package buffered implements an event dispatcher with buffering semantics
+package buffered
+
+import (
+	"context"
+	"sync"
+
+	installpb "github.com/gravitational/gravity/lib/install/proto"
+	"github.com/gravitational/gravity/lib/install/server/dispatcher"
+
+	"github.com/gravitational/trace"
+	log "github.com/sirupsen/logrus"
+)
+
+// New creates an event dispatcher with internal message buffer
+// to collect messages that it failed to send to the client.
+// The messages are resent whenever the client reconnects.
+func New() *Dispatcher {
+	ctx, cancel := context.WithCancel(context.Background())
+	d := &Dispatcher{
+		log:     log.WithField(trace.Component, "dispatcher:buffered"),
+		ctx:     ctx,
+		cancel:  cancel,
+		respC:   make(chan *installpb.ProgressResponse),
+		notifyC: make(chan *installpb.ProgressResponse),
+	}
+	d.startMessageBufferLoop()
+	return d
+}
+
+// Implements EventDispatcher
+func (r *Dispatcher) Send(event dispatcher.Event) {
+	select {
+	case r.respC <- event.AsProgressResponse():
+	case <-r.ctx.Done():
+	}
+}
+
+// Close stops the dispatcher internal processes
+// Implements EventDispatcher
+func (r *Dispatcher) Close() {
+	r.cancel()
+	r.wg.Wait()
```

Comment:
Nit: I wonder if that's a good semantics for "close" to be potentially blocking.

Reply:
It is only blocking if there's a bug in the implementation - it does not spawn goroutines it does not control that might be the cause of the block in a non-deterministic way.
I could add context to the method but I don't think it will make the API simpler or more robust.